src/topology/class.c

```diff
@@ -148,6 +148,27 @@ static int tplg_parse_class_constraints(snd_tplg_t *tplg, snd_config_t *cfg,
 	return 0;
 }
 
+/* check if mandatory and immutable attributes have been provided a value */
+static bool tplg_class_attribute_sanity_check(struct tplg_class *class)
+{
+	struct list_head *pos;
+
+	list_for_each(pos, &class->attribute_list) {
+		struct tplg_attribute *attr = list_entry(pos, struct tplg_attribute, list);
+
+		/* if tuple is mandatory and immutable, it must have a value */
+		if ((attr->constraint.mask & TPLG_CLASS_ATTRIBUTE_MASK_MANDATORY) &&
+		    (attr->constraint.mask & TPLG_CLASS_ATTRIBUTE_MASK_IMMUTABLE) &&
+		    !attr->found) {
+			SNDERR("Mandatory immutable attribute '%s' not provide for class '%s'",
+			       attr->name, class->name);
+			return false;
```

Comment:
btw, do we have the full attribute name here i.e. so we can figure out th elocation of the error ?

Reply:
if not, we can add a TODO: to these.